### ISSUE TITLE
If POST body is not read, it is not flushed.

### ISSUE THREAD

**Issue by ghotiphud** [iron/iron#444]
While testing Iron, I've found that if I POST some data, and quickly click another link the second one will return 404.  After enabling Logger, I'm seeing the method of the second request looks like `foo=1&bar=2GET`.  Fixing my POST handler to actually read the data (via urlencoded) makes this issue disappear.

**Comment by fuchsnj** [comment]
I have the same issue. This causes the HTTP method on the next request to parse incorrectly, and looks like this:
```
request method: Options
request method: Post
request method: Options
request method: Extension("{\"password\":\"\"}POST")
```

See the issue causing this in hyper I opened a while ago: https://github.com/hyperium/hyper/issues/652
And it looks like a fixed was merged recently: https://github.com/hyperium/hyper/issues/309

**Comment by untitaker** [comment]
There were also a few releases inbetween, so `cargo update` probably should fix this issue automatically. However, I wonder if we should explicitly require at least the `0.9.x` release that brought this fix.

**Comment by untitaker** [comment]
Bump, does upgrading hyper fix this?

**Comment by Ryanl92** [comment]
Hi, I've just run into this problem as well. Here is a minimal working example of the problem, if you load this in your browser. (Well in Firefox 50, linux at least). The linked ticket in hyper suggest the problem is resolved in 1.10 though.

https://gist.github.com/Ryanl92/f973f79d8800277ac1013d257844a101

**Comment by matthauck** [comment]
I'm able to repro this still with the latest versions of stuff: https://gist.github.com/matthauck/58756f0be5907673ded2254fba59f07a

**Comment by untitaker** [comment]
Please post your Cargo.lock.

**Comment by matthauck** [comment]
Ah yes, that would be helpful. Posted. Interesting there are two versions of hyper listed...

**Comment by untitaker** [comment]
Ok, I assume this is fixed in the async branch of hyper and will be fixed when that is released.

**Comment by matthauck** [comment]
`/shrug`  Hopefully...?

Thoughts on implementing a workaround for now? It seems to me pretty common that a request is not read (i.e. early return from middleware because some session header/cookie was not present.). Unclear what the timeline is for the release of that async branch of hyper.

Maybe closing the connection when there is still data left to read after a request is done? I've implemented an after chain that reads data left on requests as a `link_after`, and it seems to have fixed it where I was running into it. This is not ideal since it could end up reading data forever (the reason https://github.com/hyperium/hyper/commit/5f432851882c6f5e2dbf4efea464978bc4d88700 was reverted), and I think it doesn't solve the early return case.

**Comment by untitaker** [comment]
I'm unsure about just reading in all the data. This seems like it would have severe perf impact in some situations (like rejecting massive file upload due to failed precondition or missing permissions)

**Comment by sfackler** [comment]
I've added an after middleware that reads the remaining request body to a project I'm working on for this kind of thing. Hyper should probably have some logic to dispose of connections that haven't had their bodies consumed though - cc @seanmonstar

**Comment by seanmonstar** [comment]
Yea, it's a known issue in sync hyper (0.10 and before). The fix would be fairly difficult to do, and resources aren't really dedicated at the 0.10.x branch anymore.

The fix in async was basically to just close the connection if the full body wasn't read. It's the sanest default.

As for the bug, if I recall correctly, the only thing really happens is an error log line is generated and then the connection is dropped, when it tries to parse the remaining body as a new request.

**Comment by matthauck** [comment]
@untitaker: yeah, not advocating reading in all the data. That is not a good solution for the framework, just a middleware workaround for me I've added. I was suggesting iron could maybe drop the connection if it was not read all the way -- the `Request` object does not expose a way to do this, but thought maybe iron had a way to do this internally.

@seanmonstar: interesting. the way this manifests in iron is that the remaining data from the original request becomes part of the method for the following request. See output in [the gist above](https://gist.github.com/matthauck/58756f0be5907673ded2254fba59f07a#file-log-txt). Maybe there is actually a bug in how iron is handling this beyond the known issue in sync hyper?

**Comment by sfackler** [comment]
When a connection with an unconsumed body is recycled for keepalive, Hyper's internals will immediately try to read a new request for it, get the remaining request body, fail to parse it as an HTTP header, and drop the connection. Normally this is fine, but if the client *immediately* tries to reuse the connection for another request, that can happen before it's notified that the connection dropped.

**Comment by matthauck** [comment]
@sfackler: Are you referring to new or old behavior in hyper?

**Comment by sfackler** [comment]
This is the behavior of current hyper (0.10)
